teamcity: failed tests on master: testrace/TestMonotonicInserts

The following tests appear to have failed:

[#687614](https://teamcity.cockroachdb.com/viewLog.html?buildId=687614):


```
--- FAIL: testrace/TestMonotonicInserts/distsql=off (0.000s)
Race detected!

------- Stdout: -------
==================
Read at 0x00c42082e070 by goroutine 531:
  reflect.Value.Int()
      /usr/local/go/src/reflect/value.go:929 +0x13d
  encoding/gob.encInt()
      /usr/local/go/src/encoding/gob/encode.go:188 +0x50
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.(*Encoder).encode()
      /usr/local/go/src/encoding/gob/encode.go:701 +0x24b
  encoding/gob.(*Encoder).EncodeValue()
      /usr/local/go/src/encoding/gob/encoder.go:250 +0x5eb
  encoding/gob.(*Encoder).Encode()
      /usr/local/go/src/encoding/gob/encoder.go:175 +0x72
  github.com/cockroachdb/cockroach/pkg/kv.GRPCTransportFactory.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport_race.go:83 +0x18e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264 +0xf3

Previous write at 0x00c42082e070 by goroutine 672:
  github.com/cockroachdb/cockroach/pkg/roachpb.(*Transaction).Update()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/hlc/timestamp.go:106 +0xbfc
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1126 +0x973
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeatLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:967 +0x3f8
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1333 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264 +0xf3

Goroutine 531 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:259 +0x14c
  github.com/cockroachdb/cockroach/pkg/kv.GRPCTransportFactory()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport_race.go:61 +0x2f7
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1263 +0x14e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:393 +0x2be
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:460 +0x265
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1053 +0x3dd
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:738 +0x1a05
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:655 +0x565
  github.com/cockroachdb/cockroach/pkg/internal/client.lookupRangeFwdScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:339 +0x4e7
  github.com/cockroachdb/cockroach/pkg/internal/client.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:202 +0xb0e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:299 +0x104
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:434 +0x42d
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:310 +0x167
  github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:118 +0x4c

Goroutine 672 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:259 +0x14c
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1331 +0x673
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:494 +0x37f
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:553 +0x106
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:973 +0x66e
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:500 +0x96
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:468 +0x13d
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:500 +0xfd
  github.com/cockroachdb/cockroach/pkg/sql.(*tableWriterBase).finalize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter.go:162 +0x176
  github.com/cockroachdb/cockroach/pkg/sql.(*tableInserter).finalize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter_insert.go:57 +0x75
  github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).BatchedNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:458 +0x3ad
  github.com/cockroachdb/cockroach/pkg/sql.(*serializeNode).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:123 +0xe6
  github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:335 +0xaa
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).forEachRow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:764 +0x7c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithLocalEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:709 +0x2fe
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:657 +0xbae
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:398 +0x108f
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:92 +0x46e
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1108 +0x1ebc
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:401 +0x1e0
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:266 +0x1a9
==================
--- FAIL: testrace/TestMonotonicInserts (0.000s)
Test ended in panic.

--- FAIL: testrace/TestMonotonicInserts/distsql=off (0.000s)
Race detected!

------- Stdout: -------
==================
Read at 0x00c42082e070 by goroutine 531:
  reflect.Value.Int()
      /usr/local/go/src/reflect/value.go:929 +0x13d
  encoding/gob.encInt()
      /usr/local/go/src/encoding/gob/encode.go:188 +0x50
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.encOpFor.func4()
      /usr/local/go/src/encoding/gob/encode.go:581 +0xea
  encoding/gob.(*Encoder).encodeStruct()
      /usr/local/go/src/encoding/gob/encode.go:328 +0x3ec
  encoding/gob.(*Encoder).encode()
      /usr/local/go/src/encoding/gob/encode.go:701 +0x24b
  encoding/gob.(*Encoder).EncodeValue()
      /usr/local/go/src/encoding/gob/encoder.go:250 +0x5eb
  encoding/gob.(*Encoder).Encode()
      /usr/local/go/src/encoding/gob/encoder.go:175 +0x72
  github.com/cockroachdb/cockroach/pkg/kv.GRPCTransportFactory.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport_race.go:83 +0x18e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264 +0xf3

Previous write at 0x00c42082e070 by goroutine 672:
  github.com/cockroachdb/cockroach/pkg/roachpb.(*Transaction).Update()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/hlc/timestamp.go:106 +0xbfc
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeat()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1126 +0x973
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeatLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:967 +0x3f8
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1333 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264 +0xf3

Goroutine 531 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:259 +0x14c
  github.com/cockroachdb/cockroach/pkg/kv.GRPCTransportFactory()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport_race.go:61 +0x2f7
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1263 +0x14e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:393 +0x2be
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:460 +0x265
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1053 +0x3dd
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:738 +0x1a05
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:655 +0x565
  github.com/cockroachdb/cockroach/pkg/internal/client.lookupRangeFwdScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:339 +0x4e7
  github.com/cockroachdb/cockroach/pkg/internal/client.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:202 +0xb0e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:299 +0x104
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:434 +0x42d
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:310 +0x167
  github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:118 +0x4c

Goroutine 672 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:259 +0x14c
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1331 +0x673
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:494 +0x37f
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:553 +0x106
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:973 +0x66e
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:500 +0x96
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:468 +0x13d
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:500 +0xfd
  github.com/cockroachdb/cockroach/pkg/sql.(*tableWriterBase).finalize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter.go:162 +0x176
  github.com/cockroachdb/cockroach/pkg/sql.(*tableInserter).finalize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter_insert.go:57 +0x75
  github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).BatchedNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:458 +0x3ad
  github.com/cockroachdb/cockroach/pkg/sql.(*serializeNode).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:123 +0xe6
  github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:335 +0xaa
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).forEachRow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:764 +0x7c
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithLocalEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:709 +0x2fe
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:657 +0xbae
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:398 +0x108f
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:92 +0x46e
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1108 +0x1ebc
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:401 +0x1e0
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:266 +0x1a9
==================
--- FAIL: testrace/TestMonotonicInserts (12.890s)
Test ended in panic.


```

Please assign, take a look and update the issue accordingly.

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -718,6 +718,12 @@ type response struct {
 func (ds *DistSender) divideAndSendBatchToRanges(
 	ctx context.Context, ba roachpb.BatchRequest, rs roachpb.RSpan, batchIdx int,
 ) (br *roachpb.BatchResponse, pErr *roachpb.Error) {
+	// Clone the BatchRequest's transaction so that future mutations to the
+	// proto don't affect the proto in this batch.
+	if ba.Txn != nil {
+		txnCopy := *ba.Txn
+		ba.Txn = &txnCopy
+	}
 	// Get initial seek key depending on direction of iteration.
 	var scanDir ScanDirection
 	var seekKey roachpb.RKey
@@ -867,14 +873,7 @@ func (ds *DistSender) divideAndSendBatchToRanges(
 		// batch RPCs, send asynchronously.
 		if canParallelize && !lastRange && ds.rpcContext != nil &&
 			ds.sendPartialBatchAsync(ctx, ba, rs, ri.Desc(), ri.Token(), batchIdx, responseCh) {
-			// Note that we pass the batch request by value to the parallel
-			// goroutine to avoid using the cloned txn.
-
-			// Clone the txn to preserve the current txn sequence for the async call.
-			if ba.Txn != nil {
-				txnClone := ba.Txn.Clone()
-				ba.Txn = &txnClone
-			}
+			// Sent the batch asynchronously.
 		} else {
 			resp := ds.sendPartialBatch(ctx, ba, rs, ri.Desc(), ri.Token(), batchIdx, true /* needsTruncate */)
 			responseCh <- resp

